aiohttp/connector.py

```diff
@@ -1085,6 +1085,10 @@ async def _create_direct_connection(
 
         host = req.url.raw_host
         assert host is not None
+        # Replace multiple trailing dots with a single one.
+        # See https://github.com/aio-libs/aiohttp/pull/7364.
+        while host.endswith(".."):
+            host = host[:-1]
```

Comment:
Multiple string assignments and recreations seem rather expensive. How about:
```suggestion
        if host.endswith("..") and (trailing_dots_cutoff := -2):
            while test_host[trailing_dots_cutoff - 1] == ".":
                trailing_dots_cutoff -= 1
            host = host[:trailing_dots_cutoff]
```

Reply:
Or just `host = host.rstrip(".") + "."`?

Reply:
```
> python3 -m timeit 'host = "foo.bar....."' 'host = host.rstrip(".") + "."'
2000000 loops, best of 5: 144 nsec per loop
> python3 -m timeit 'host = "foo.bar....."' 'while host.endswith(".."):
 host = host[:-1]'

500000 loops, best of 5: 702 nsec per loop
> python3 -m timeit 'host = "foo.bar....."' 'if host.endswith("..") and (trailing_dots_cutoff := 2):
 while host[trailing_dots_cutoff - 1] == ".":
  trailing_dots_cutoff -= 1
 host = host[:trailing_dots_cutoff]'
1000000 loops, best of 5: 217 nsec per loop
```

Reply:
Wow, that's even better! I also tested your variant with an f-string instead of concatenation, and it's a bit slower. So let's go with your suggestion:
```suggestion
        host = host.rstrip(".") + "."
```

Reply:
`host = host.rstrip(".") + "."` adds a dot at the end if it was not present, so we can't use it directly. (Names without the dot at the end are resolved using the domain search path, with dot at the end the resolver does not use search path). With if block it is a bit slower:

```
> python3 -m timeit 'host = "foo.bar....."' 'host = host.rstrip(".") + "."'
5000000 loops, best of 5: 86.7 nsec per loop
> python3 -m timeit 'host = "foo.bar....."' 'if host.endswith("."): host = host.rstrip(".") + "."'
1000000 loops, best of 5: 189 nsec per loop
> python3 -m timeit 'host = "foo.bar....."' 'host = host[:len(host.rstrip("."))+1]'
2000000 loops, best of 5: 134 nsec per loop

> python3 -m timeit 'host = "foo.bar"' 'host = host.rstrip(".") + "."'
5000000 loops, best of 5: 54.4 nsec per loop
> python3 -m timeit 'host = "foo.bar"' 'if host.endswith("."): host = host.rstrip(".") + "."'
5000000 loops, best of 5: 72.7 nsec per loop
> python3 -m timeit 'host = "foo.bar"' 'host = host[:len(host.rstrip("."))+1]'
5000000 loops, best of 5: 74.9 nsec per loop
```

I opted for `if host.endswith("."): host = host.rstrip(".") + "."` anyway because it is more explicit about the missing dot case. Please let me know if you want to change it.

Reply:
Right, I thought I included that check. Thanks for spotting that!